module/zfs/vdev_raidz.c

```diff
@@ -141,6 +141,181 @@
+ *    two reasons. First, we will have to read from all disks in the RAIDZ
+ *    VDEV, rather than being able to skip the children that contain only
+ *    parity of this block (because the parity and data of a single block are
+ *    now spread out across all the disks).  Second, in most cases there will
+ *    be an extra bcopy, needed to rearrange the data back to its original
+ *    layout in memory.
+ *
+ * == Scratch Area ==
+ *
+ * As we copy the block data, we can only progress to the point that writes
+ * will not overlap with blocks whose progress has not yet been recorded on
+ * disk.  Since partially-copied rows are always read from the old location,
+ * we need to stop one row before the sector-wise overlap, to prevent any
+ * row-wise overlap. Initially this would limit us to copying one sector at
+ * a time. The amount we can safely copy is known as the chunk size.
+ *
+ * Ideally we want to copy at least 2 * (new_width)^2 so that we have a
+ * separation of 2*(new_width+1) and a chunk size of new_width+2. With the max
+ * RAIDZ width of 255 and 4K sectors this would be 2MB per disk. In practice
+ * the widths will likely be single digits so we can get a substantial chuck
+ * size using only a few MB of scratch per disk.
+ *
+ * To speed up the initial copy, we use a scratch area that is persisted to
+ * disk which holds a large amount of reflowed state. We can always read the
+ * partially written stripes when a disk fails or the copy is interrupted
+ * (crash) during the initial copying phase and also get past a small chunk
+ * size restriction.  At a minimum, the scratch space must be large enough to
+ * get us to the point that one row does not overlap itself when moved
+ * (i.e new_width^2).  But going larger is even better. We use the 3.5 MiB
+ * reserved "boot" space that resides after the ZFS disk labels as our scratch
+ * space to handle overwriting the initial part of the VDEV.
+ *
+ *	0     256K   512K                    4M
+ *	+------+------+-----------------------+-----------------------------
+ *	| VDEV | VDEV |   Boot Block (3.5M)   |  Allocatable space ...
+ *	|  L0  |  L1  |       Reserved        |     (Metaslabs)
+ *	+------+------+-----------------------+-------------------------------
+ *                        Scratch Area
+ *
+ * == Reflow Progress Updates ==
+ * After the initial scratch-based reflow, the expansion process works
+ * similarly to device removal. We create a new open context thread whichi
+ * reflows the data, and periodically kicks off sync tasks to update logical
+ * state. In this case, state is the committed progress (offset of next data
+ * to copy). We need to persist the completed offset on disk, so that if we
+ * crash we know which format each VDEV offset is in.
+ *
+ * == Time Dependent Geometry ==
+ *
+ * In RAIDZ, blocks are read from disk in a column by column fashion. For a
+ * multi-row block, the second sector is in the first column not in the 2nd
```

Comment:
FreeBSD's doc style checker (igor) complained about "2nd" and I tend to agree, given "first" and "second" elsewhere in the sentence.

(It also complained that a comma should follow `i.e.`, but that may be overly pedantic.)

Reply:
Thanks @emaste. fixed in latest commit.